smokerand.1
.TH smokerand 1  "September 14, 2026" "version 0.51" "USER COMMANDS"
.SH NAME
smokerand \- a test suite for pseudorandom numbers generators
.SH SYNOPSIS
.B smokerand
\fIbattery\fR \fIgenerator_lib\fR [\fIkeys\fR]
.SH DESCRIPTION
.PP
SmokeRand is a cross-platform set of tests for pseudorandom number generators
(PRNGs). Tested generators should return either 32\-bit or 64\-bit unsigned
uniformly distributed unsigned integers. They can be connected to SmokeRand
either through stdin/stdout or by means of C API as dynamic libraries.
It has the next important features:
.IP \[bu]
Direct support of 64-bit generators without taking upper/lower parts,
interleaving etc.
.IP \[bu]
Analysis of the lowest bits of generators output that improves sensitivity
of tests, e.g. linear complexity, matrix rank or birthday spacings test.
.IP \[bu]
Multithreading support by means of POSIX threads or WinAPI threads.
In this case PRNG must be supplied as a dynamic library.
.IP \[bu]
A minimalistic set of fast, relatively simple and mostly well known
statistical tests that cover the most common artefacts of popular PRNGs.
Users can either apply several predefined batteries or write custom
batteries by means of a human-readable text files.
.PP
The main idea behind SmokeRand is development of statistical tests batteries
that would take into account weak spots of TestU01 (testing of lower bits,
64-bit generators, absence of some newer criteria), PractRand (absence of
birthday spacings and linear complexity tests) and dieharder (relatively low
sensitivity of included tests). The resulting program cannot replace them
but its batteries are diverse enough to be used as a "smoke test" for PRNGs.
However it easily detects statistical flaws in truncated 128-bit LCGs with
power of 2 modulo, additive/subtractive lagged Fibonacci generators with large
lags up to 44497, xoroshiro128+, Mersenne Twister, TinyMT32/64, SplitMix
and even DES in CTR mode.
.PP
All statistical tests included in SmokeRand batteries return p-value. Their
interpretation is rather simple and based on the next intervals:
.IP \[bu]
\fBOk\fR: in the (0.001 and 0.999) interval;
.IP \[bu]
\fBSUSPICIOUS\fR: in the [1e\-10; 0.001] and [0.999; 1 \- 1e-10] intervals;
.IP \[bu]
\fBFAIL\fR: in the [0; 1e\-10) and (1 \- 1e\-10; 1] intervals.
.PP
Only systematic \fBSUSPICIOUS\fR or \fBFAIL\fR results must be interpreted
as detected statistical flaw of the tested pseudorandom number generator.
Such threshold values may seem unusually strict but allow to significantly
reduce false positives. Tests sensitivity can be easily increased by using
larger samples without altering critical p-values. Such approach is used in
dieharder, TestU01, and PractRand test suites.
.PP
\fBWARNING!\fR The SmokeRand test batteries can detect flaws in many popular
pseudorandom number generators but they \fBcannot guarantee a high quality\fR
of generator that passed them, \fBespecially for cryptography\fR. Consider
them just as a "smoke test" for searching obvious defects in pseudorandom number
generators.
.PP
A state-of-art general purpose PRNG must be cryptographically secure, i.e. its
quality should be analysed not only by empirical test but by cryptoanalytic
techniques. Of course, it should be used correctly, i.e. having period at least
2**64, pass not only the common statistical tests but also the 64-bit collision
test, not be reseeded during the sample generation if reproducible simulation
is desired etc.
.SS Automatic quality assessment
.PP
SmokeRand makes an automatic quality assessment in the form of numerical grade
from 0 (the lowest quality) to 4 (the highest quality). The grades are
interpreted the next way:
.IP \[bu]
\fB4.0\fR \- \fBgood\fR generator, probably may be used as a general purpose PRNG.
.IP \[bu]
\fB3.0\-3.99\fR \- \fBsome issues\fR have been detected, usage for any computational
task requires an extra analysis.
.IP \[bu]
\fB2.0\-2.99\fR \- \fBflawed\fR generator. Cannot be recommended as a general purpose
generator but detected flaws will probably cause problems only in some specific
conditions.
.IP \[bu]
\fB1.0\-1.99\fR \- \fBbad\fR generator. Detected biases are serious and can easily
distort results of Monte-Carlo simulations. But it may be suitable for e.g.
programming games like TETRIS or generating some non-repeating values.
.IP \[bu]
\fB<1.0\fR \- \fBvery bad\fR generator. Can be used only if its replacement
to a simple counter (0,1,2,3,....) would be totally acceptable.
.PP
The \fBexcellent\fR grade is intentionally avoided because such grading is
impossible without thorough cryptoanalysis that cannot be fully automated and
represented as a battery of statistical tests.
.PP
The grading algorithm is rather simple and based on a penalty system. An initial
grade is 4.0 and some penalty is subtracted from it for an every failed test.
The used penalties for failures:
.IP \[bu]
4.0 for frequency test and gap test.
.IP \[bu]
3.0 for birthday spacings and collision over tests.
.IP \[bu]
2.0 for Hamming weights based test, gap16_count0, mod3 and sumcollector tests.
.IP \[bu]
1.0 for birthday spacings test with decimation.
.IP \[bu]
0.25 for a matrix rank and linear complexity tests. If e.g. three test of this
kind are failed \- 0.75 is subtracted from the grade.
.PP
If the resulting grade is less than zero then it is converted to zero.
Of course, this scale is very crude, partially subjective and probably rather
strict. It is aimed on distinction between e.g. low linear complexity of some
bits (important only in some specific cases) and failure of gap test (can
easily distort a Monte-Carlo simulation results).
.PP
If failures for the \fBcoll64dec\fR or \fBfreq\fR extra batteries were observed
then the 1.0 penalty value should be subtracted from the final grade for each
failure. E.g. SplitMix and RC4 will be graded as 3.0 instead of 4.0.
.PP
Examples of grading for some generators:
.IP \[bu]
\fBgood\fR: AES128, ChaCha12, xoroshiro128++, KISS99, MWC64X.
.IP \[bu]
\fBsome issues\fR: Mersenne Twister, 128-bit LCG, xoroshiro128+,
SplitMix, RC4, DES-CTR.
.IP \[bu]
\fBflawed\fR: flea32x1, taus88, WELL1024a.
.IP \[bu]
\fBbad\fR, \fBvery bad\fR: RANDU, 32-bit and 64-bit LCGs with power of 2 modulo,
xorwow, SWB, ChaCha12 with 32-bit counter, LFIB(607,273,+)
.PP
Grouping RANDU together with xorwow and additive lagged Fibonacci PRNGs
may be seen as too pessimistic quality estimation of some generators. But
it is partially based on the idea that usage of non-cryptographic generators
that even pass all tests (have 4.0 final grade) is an intentional downgrade
in quality, bithack that requires a special justification.
.SS Custom batteries
.PP
Custom batteries are kept inside a simple human-readable text files. It supports
single line comments beginning with \# and blank lines. Each non-blank line
contains description of a single test. It has the next format:
.PP

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/)

## [0.52]

### Added

- gap test now checks individual frequencies too, i.e. compares distritutions
  for each `|Oi - Ei|/Ei**0.5` with the theoretical one. It improves its 
  sensitivity to the additive/subtractive lagged Fibonacci PRNGs and SWB
  (subtract-with-borrow) PRNGs.

### Bugfix

- Threads abstraction layer refinement (some subtle data races and pthreads
  mutexes double initialization problem fixed).
- Lua scripts for Ninja/Watcom generation: an issue with the absent/misspelled
  `src/lfsr_period_factors.c` was fixed.

## [0.51] 2026-09-14

### Added

- `lfsr_bm` battery that uses Berlekamp-Massey algorithm to restore the
  characteristic polynomial. Can work with PRNGs with circle buffers/pointers
  etc. (e.g. `lrnd64_255`, `tt800`, `well1024a` etc), can use `stdin`/`stdout`
  interface. Cannot work if nonlinear scramblers are present. This battery
  also supports PRNGs based on Mersenne primes.
- `full` battery: new modifications of the birthday spacings test with expected
  number collison of 256. They are rather fast but allow to catch `grube_mrg2`
  generator.
- birthday spacings test: `lambda` (expected number of collisions) can be now
  set by a user.
- gap test: now it simultaneously applied to the lowest and highest bits,
  it is equivalent to the subtests: for raw values and for values with reversed
  bits order. Useful for `gap_inv8` test for a simplified version of Sirius64
  (with an extra round without shift).
- `hamming_distr` test now contains subtests for the lowest bit. Useful for
  catching flaws in `tf0_64`, `swbw`, `sirius64_bad`.
- `melg4253` generator was added.
- `swbmwc32`, `swbmwc64`, `swblux64`: simplification of the initialization
  procedure, now it is based on the `expand_seed64_to_u64` function and requires
  only one 64-bit seed per generator.
- `swb64sc`: a new experimental PRNG based on 64-bit subtract-with-borrow
  algorithm with a non-linear output function.
- `swb64dec`: a new experimental PRNG based on 53-bit subtract-with-borrow
  algorithm with decimation.
- `xorrot512` and `xorrot512mrt` experimental PRNGs by A.L. Voskov.
- `yasmarang` generator was added.
- `test_lfsr_period.c`: `xoshiro512++` was added as a test example.
- `include/coredefs.h`: `swb_u64` function wad added to provide a portable and
  reliable support of subtract with borrow for unsigned 64-bit integers.

### Bugfix

- Typo in `kiss93` was fixed (introduced during some recent refactoring).
- Some tables of prime factors in `lfsr_period.c` didn't have `LFSR_EXPS_END`
  terminator (undefined behaviour!)
- `hamming_distr_calc_zemp`: now it takes into account the number of tuples
  with the maximal Hamming weight. May be detectable only for 32-bit tuples.
- `swbmwc64`, `swblux64`: now uses `swb_u64` to prevent rare and subtle errors.
- Build fixes for MSVC and Open Watcom.

### Changes

- Refactoring: more compact assertions for malloc in some places.
- Some improvements in `src/threads_intf.c`: improved error processing and less
  fragile `ThreadObj` methods.
- Refactoring: `m - 1` prime cofactors for `src/lfsr_period.c` are now separated
  into `src/lfsr_period_factors.c` and `include/smokerand/lfsr_period_factors.h`.

## [0.50] 2026-09-01

### Added


apps/smokerand.c
/**
 * @file smokerand.c
 * @brief SmokeRand command line interface.
 * @copyright
 * (c) 2024-2026 Alexey L. Voskov, Lomonosov Moscow State University.
 * alvoskov@gmail.com
 *
 * This software is licensed under the MIT license.
 */
#include "smokerand_core.h"
#include "smokerand_bat.h"
#include "smokerand/fileio.h"
#include "smokerand/threads_intf.h"
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <string.h>
#include <float.h>

void print_help(void)
{
    static const char help_str[] = 
    "SmokeRand: a test suite for pseudorandom number generators\n"
    "(C) 2024-2026 Alexey L. Voskov\n\n"
    "Usage: smokerand [battery] generator_lib [keys]\n"
    "battery: battery name; supported batteries:\n"
    "  General purpose batteries\n"
    "  - express    Express battery (32-64 MiB of data)\n"
    "  - brief      Fast battery (64-128 GiB of data)\n"
    "  - default    Slower but more sensitive battery (128-256 GiB of data)\n"
    "  - full       The slowest battery (1-2 TiB of data)\n"
    "  Special batteries\n"
    "  - coll64dec  64-bit collision test with decimation (former `birthday`).\n"
    "  - freq       8-bit and 16-bit words frequency adaptive tests.\n"  
    "  - ising      Ising model based tests: Wolff and Metropolis algorithms.\n"
    "  - lfsr       LFSR generators maximal period automatic verifier.\n"
    "  - f=filename Load a custom battery from the text config file.\n"
    "  - s=filename Load a custom battery implemented as a shared library.\n"
    "  Special modes\n"
    "  - help       Print a built-in PRNG help (if available).\n"
    "  - selftest   Runs PRNG internal self-test (if available).\n"
    "  - speed      Measure speed of the generator\n"
    "  - stdout     Sends PRNG output to stdout in the binary form.\n"
    "  - stdoutfl   Sends PRNG output to stdout in the floating point form.\n"
    "  - stdoutflx  Sends PRNG output to stdout in the floating point form\n"
    "               using an improved algorithm of generation\n"
    "generator_lib: name of dynamic library with PRNG or special mode name.\n"
    "  Special modes names:\n"
    "  - stdin32, stdin64  Get random sequence from stdin\n"
    "  - list              Print list of tests in the battery\n"
    "Optional keys\n"
    "  --filter=name Apply pre-defined filter to the generator output\n"
    "    reverse-bits   Reverse bits in the generator output\n"
    "    interleaved32  Process 64-bit generator output as interleaving 32-bit words\n"
    "    high32, low32  Analyse higher/lower 32 bits of 64-bit generator\n"
    "    uint31, uint63 Autocomplete 31/63-bit generator (the lowest bit is 0)\n"
    "                   to the 32/64-bit one using Murmur3 mixer for the lowest bit\n"
    "                   generation.\n"
    "  --maxlen_log2=n  Limit the binary output to the 2^n floats (12 <= n <= 63)\n"
    "  --report-brief Show only failures in the report\n"
    "  --seed=data Use the user supplied string (data) as a seed\n"
    "  --testid=id     Run only the test with the given numeric id\n"
    "  --testname=name Run only the test with the given name\n"
    "  --nthreads  Run battery in multithreaded mode (default number of threads)\n"
    "  --threads=n Run battery in multithreaded mode using n threads\n"
    "\n";

    printf(help_str);
}


typedef enum {
    FILTER_NONE,
    FILTER_REVERSE_BITS,
    FILTER_INTERLEAVED32,
    FILTER_HIGH32,
    FILTER_LOW32,
    FILTER_UINT31,
    FILTER_UINT63,
    FILTER_UNKNOWN
} GeneratorFilter;


GeneratorFilter GeneratorFilter_from_name(const char *name)
{
    if (!strcmp("reverse-bits", name)) {
        return FILTER_REVERSE_BITS;
    } else if (!strcmp("interleaved32", name)) {
        return FILTER_INTERLEAVED32;
    } else if (!strcmp("high32", name)) {
        return FILTER_HIGH32;
    } else if (!strcmp("low32", name)) {
        return FILTER_LOW32;
    } else if (!strcmp("uint31", name)) {
        return FILTER_UINT31;
    } else if (!strcmp("uint63", name)) {
        return FILTER_UINT63;
    } else {
        return FILTER_UNKNOWN;
    }
}

typedef struct {
    unsigned int nthreads; ///< From `--threads` or `--nthreads` keys.
    unsigned int nthreads_from_seed; ///< From base64 seed (`seed=_` key)
    unsigned int testid; ///< Test identifier obtained from the `--testid` key
    const char *testname; ///< Test name obtained from the `--testname` key
    const char *bat_param;
    unsigned int maxlen_log2; ///< log2(len) for stdout length in bytes
    GeneratorFilter filter;
    ReportType report_type;
} SmokeRandSettings;

/**
 * @brief Returns a default number of threads suitable for the current
 * hardware configuration.
 * @details General principles:
 *
 * 1) Maximal number of threads is the number of CPU cores.
 * 2) For 32-bit systems - not more than 2 threads.
 * 3) If number of cores is more than 4 then leave one unloaded core.
 */
static unsigned int get_default_nthreads(unsigned int *ncores)
{
    unsigned int nthreads = get_cpu_numcores();
    if (ncores != NULL) {
        *ncores = nthreads;
    }
    if (sizeof(size_t) == 4 * sizeof(char) && nthreads > 2) {
        nthreads = 2;
    }
    if (nthreads > 4)
        nthreads--;
    return nthreads;
}


typedef BatteryExitCode (*ArgValueCallback) (SmokeRandSettings *obj, const char *argvalue);


#define DEFINE_NUMARG_CALLBACK(field, argname, condition) \
static BatteryExitCode field##_callback(SmokeRandSettings *obj, const char *argvalue) \
{ \
    int argval = atoi(argvalue); \
    if (condition) { \
        obj->field = (unsigned int) argval; \
        return BATTERY_PASSED; \
    } else { \
        fprintf(stderr, "Invalid value of argument '%s'\n", argname); \
        return BATTERY_ERROR; \
    } \
}

typedef struct {
    const char *name;
    ArgValueCallback callback;
} ArgumentEntry;


static BatteryExitCode process_argument(SmokeRandSettings *obj,
    const ArgumentEntry *args,
    const char *argname, const char *argvalue)
{
    for (const ArgumentEntry *e = args; e->name != NULL; e++) {
        if (!strcmp(e->name, argname)) {
            if (e->callback(obj, argvalue) == BATTERY_PASSED) {
                return BATTERY_PASSED;
            } else {
                return BATTERY_ERROR;
            }
        }
    }
    return BATTERY_FAILED;
}



DEFINE_NUMARG_CALLBACK(nthreads, "nthreads", argval > 0)
DEFINE_NUMARG_CALLBACK(testid, "testid", argval > 0)
DEFINE_NUMARG_CALLBACK(maxlen_log2, "maxlen_log2", 12 <= argval || argval <= 63)


static BatteryExitCode SmokeRandSettings_numarg_load(SmokeRandSettings *obj,
    const char *argname, const char *argvalue)
{
    static const ArgumentEntry args[] = {
        {"nthreads", nthreads_callback},
        {"testid",   testid_callback},
        {"maxlen_log2", maxlen_log2_callback},
        {NULL, NULL}
    };
    return process_argument(obj, args, argname, argvalue);
}

static inline int char_to_hex_digit(char c)
{
    if ('0' <= c && c <= '9') {
        return (int) (c - '0');
    } else if ('A' <= c && c <= 'F') {
        return (int) (c - 'A') + 10;
    } else if ('a' <= c && c <= 'f') {
        return (int) (c - 'a') + 10;
    } else {
        return -1;
    }
}

static BatteryExitCode SmokeRandSettings_txtarg_load(SmokeRandSettings *obj,
    const char *argname, const char *argvalue)
{
    if (!strcmp(argname, "param")) {
        set_cmd_param(argvalue);
        return BATTERY_PASSED;
    } else if (!strcmp(argname, "batparam")) {
        obj->bat_param = argvalue;
        return BATTERY_PASSED;
    } else if (!strcmp(argname, "filter")) {
        obj->filter = GeneratorFilter_from_name(argvalue);
        if (obj->filter == FILTER_UNKNOWN) {
            fprintf(stderr, "Unknown filter %s\n", argvalue);
            return BATTERY_ERROR;
        } else {
            return BATTERY_PASSED;
        }
    } else if (!strcmp(argname, "seed")) {
        if (strlen(argvalue) > 5 && argvalue[0] == '_') {            
            if (argvalue[3] != '_') {
                fprintf(stderr, "Invalid format of _ seed\n");
                return BATTERY_ERROR;
            }
            int d1 = char_to_hex_digit(argvalue[1]);
            int d0 = char_to_hex_digit(argvalue[2]);
            if (d0 < 0 || d1 < 0) {
                fprintf(stderr, "Invalid format of number of threads in the _ seed\n");
                return BATTERY_ERROR;
            }
            obj->nthreads_from_seed = (unsigned int) (d1 * 16 + d0);
            if (!set_entropy_base64_seed(argvalue + 4)) {
                return BATTERY_ERROR;
            }
        } else {            
            set_entropy_textseed(argvalue);
        }
        return BATTERY_PASSED;
    } else if (!strcmp(argname, "testname")) {
        obj->testname = argvalue;
        return BATTERY_PASSED;
    } else {
        return BATTERY_FAILED;
    }
}


#define PROCESS_ARGUMENTS_BLOCK(loader_func) \
{ \
    BatteryExitCode code = loader_func(obj, argname, eqpos + 1); \
    if (code == BATTERY_ERROR) { \
        return BATTERY_ERROR; \
    } else if (code == BATTERY_PASSED) { \
        continue; \
    } \
}


void SmokeRandSettings_init(SmokeRandSettings *obj)
{
    obj->nthreads           = 1;
    obj->testid             = TESTS_ALL;
    obj->testname           = NULL;
    obj->report_type        = REPORT_FULL;
    obj->bat_param          = NULL;
    obj->nthreads_from_seed = 0;
    obj->filter             = FILTER_NONE;
    obj->maxlen_log2        = 0;
}

/**
 * @brief Process command line arguments to extract settings.
 * @param[out] obj  Output buffer for parsed settings.
 * @param[in]  argc Number of command line arguments (from `main` function).
 * @param[in]  argv Values of command line arguments (from `main` function).
 */
BatteryExitCode SmokeRandSettings_load(SmokeRandSettings *obj, int argc, char *argv[])
{
    SmokeRandSettings_init(obj);
    for (int i = 3; i < argc; i++) {
        char argname[32];
        char *eqpos = strchr(argv[i], '=');
        size_t len = strlen(argv[i]);
        if (!strcmp(argv[i], "--threads")) {
            unsigned int ncores;
            obj->nthreads = get_default_nthreads(&ncores);
            fprintf(stderr, "%u CPU cores detected\n", ncores);
            fprintf(stderr, "%u threads would be created\n", obj->nthreads);
            continue;
        }
        if (!strcmp(argv[i], "--report-brief")) {
            obj->report_type = REPORT_BRIEF;
            continue;
        }
        if (len < 3 || (argv[i][0] != '-' || argv[i][1] != '-') || eqpos == NULL) {
            fprintf(stderr, "Argument '%s' should have --argname=argval layout\n", argv[i]);
            return BATTERY_ERROR;
        }
        ptrdiff_t name_len = (eqpos - argv[i]) - 2;
        if (name_len >= 32) name_len = 31;
        if (name_len > 0) {
            memcpy(argname, &argv[i][2], (size_t) name_len);
            argname[name_len] = '\0';
        } else {
            argname[0] = '\0';
        }
        // Text arguments
        PROCESS_ARGUMENTS_BLOCK(SmokeRandSettings_txtarg_load);
        // Numerical arguments
        PROCESS_ARGUMENTS_BLOCK(SmokeRandSettings_numarg_load);
        // Unknown argument
        fprintf(stderr, "Unknown argument '%s'\n", argname);
        return BATTERY_ERROR;

    }
    if (obj->nthreads_from_seed != 0) {
        obj->nthreads = obj->nthreads_from_seed;
    }
    return BATTERY_PASSED;
}


typedef BatteryExitCode (*BatteryCallback)(const GeneratorInfo *gen,
    const CallerAPI *intf, const BatteryOptions *opts);


typedef struct {
    const char *name;
    BatteryCallback battery;    
} BatteryEntry;


static BatteryExitCode battery_dummy(const GeneratorInfo *gen, const CallerAPI *intf,
    const BatteryOptions *opts)
{
    fprintf(stderr, "Battery 'dummy': do nothing\n");
    (void) gen; (void) intf; (void) opts;
    return BATTERY_PASSED;
}

static BatteryExitCode battery_help(const GeneratorInfo *gen, const CallerAPI *intf,
    const BatteryOptions *opts)
{
    (void) intf; (void) opts;
    if (gen->description != NULL) {
        printf("%s\n", gen->description);
        return BATTERY_PASSED;
    } else {
        printf("Built-in help not found\n");
        return BATTERY_FAILED;
    }
}

#define DEFINE_SHORT_BATTERY_ENV(battery_name) \
static BatteryExitCode battery_##battery_name##_env(const GeneratorInfo *gen, \
    const CallerAPI *intf, const BatteryOptions *opts) { \
    (void) opts; \
    return battery_##battery_name(gen, intf); \
}

//DEFINE_SHORT_BATTERY_ENV(collover64_decimated)
DEFINE_SHORT_BATTERY_ENV(blockfreq)
DEFINE_SHORT_BATTERY_ENV(self_test)
DEFINE_SHORT_BATTERY_ENV(speed)


/**
 * @brief Will load battery from dynamic library
 */
BatteryExitCode battery_shared_lib(const char *filename, const GeneratorInfo *gen,
    const CallerAPI *intf, const BatteryOptions *opts)
{
    void *lib = dlopen_wrap(filename);
    if (lib == NULL) {
        fprintf(stderr, "Cannot open the `%s` battery\n", filename);
        return BATTERY_ERROR;
    }
    BatteryCallback battery_func;
    BatteryExitCode result;
    void *fptr = dlsym_wrap(lib, "battery_func");
    memcpy(&battery_func, &fptr, sizeof(battery_func));
    if (battery_func == NULL) {
        fprintf(stderr, "Cannot find the 'battery_func' function\n");
        result = BATTERY_ERROR;
    } else {
        result = battery_func(gen, intf, opts);
    }
    dlclose_wrap(lib);
    return result;
}

/**
 * @brief Run a battery of statistical test for a given generator.
 * @param battery_name  Name of tests battery: `default`, `express`, `brief`,
 * `full`, `help`, `selftest`, `speed`, `stdout`, `freq`, 'birthday`, `ising`.
 * @param gi   PRNG to be tested.
 * @param intf Pointers to API functions.
 * @param opts Pre-parsed command line options.
 */
BatteryExitCode run_battery(const char *battery_name, GeneratorInfo *gi,
    CallerAPI *intf, SmokeRandSettings *opts)
{
    static const BatteryEntry batteries[] = {
        {"default",    battery_default},
        {"brief",      battery_brief},
        {"full",       battery_full},
        {"express",    battery_express},
        {"help",       battery_help},
        {"lfsr",       battery_lfsr_period},
        {"selftest",   battery_self_test_env},
        {"speed",      battery_speed_env},
        {"freq",       battery_blockfreq_env},
        {"birthday",   battery_collover64_decimated},
        {"coll64dec",  battery_collover64_decimated},
        {"ising",      battery_ising},
        {"unitsphere", battery_unit_sphere_volume},
        {"dummy",      battery_dummy},
        {NULL,         NULL}
    };

    if (opts->testid != TESTS_ALL && opts->testname != NULL) {
        fprintf(stderr, "testid and testname keys cannot coexist\n");
        return BATTERY_ERROR;
    }
    
    BatteryExitCode ans = BATTERY_UNKNOWN;
    BatteryOptions bat_opts;
    bat_opts.test.id     = opts->testid;
    bat_opts.test.name   = opts->testname;
    bat_opts.nthreads    = opts->nthreads;
    bat_opts.report_type = opts->report_type;
    bat_opts.param       = (opts->bat_param != NULL) ? opts->bat_param : "";


    if (strlen(battery_name) > 1 &&
        battery_name[0] == 'f' && battery_name[1] == '=') {
        if (battery_name[2] == '\0') {
            fprintf(stderr, "File name cannot be empty");
            return BATTERY_ERROR;
        }
        const char *filename = battery_name + 2;
        ans = battery_file(filename, gi, intf, &bat_opts);
    } else if (strlen(battery_name) > 1 && 
        battery_name[0] == 's' && battery_name[1] == '=') {
        if (battery_name[2] == '\0') {
            fprintf(stderr, "File name cannot be empty");
            return BATTERY_ERROR;
        }
        const char *filename = battery_name + 2;
        ans = battery_shared_lib(filename, gi, intf, &bat_opts);
    } else if (!strcmp(battery_name, "stdout")) {
        GeneratorInfo_bits_to_file(gi, intf, opts->maxlen_log2);
    } else if (!strcmp(battery_name, "stdoutfl")) {
        GeneratorInfo_floats_to_file(gi, intf, opts->maxlen_log2);
    } else if (!strcmp(battery_name, "stdoutflx")) {
        GeneratorInfo_accurate_floats_to_file(gi, intf, opts->maxlen_log2);
    } else {
        for (const BatteryEntry *entry = batteries; entry->name != NULL; entry++) {
            if (!strcmp(battery_name, entry->name)) {
                ans = entry->battery(gi, intf, &bat_opts);
                break;
            }
        }
        if (ans == BATTERY_UNKNOWN) {
            fprintf(stderr, "Unknown battery %s\n", battery_name);
        }
    }
    return ans;
}

int print_battery_info(const char *battery_name)
{
    const BatteryOptions opts = {
        .test = {.id = 0, .name = NULL},
        .nthreads = 0, .report_type = REPORT_FULL, .param = NULL
    };
    if (!strcmp(battery_name, "express")) {
        battery_express(NULL, NULL, &opts);
    } else if (!strcmp(battery_name, "default")) {
        battery_default(NULL, NULL, &opts);
    } else if (!strcmp(battery_name, "brief")) {
        battery_brief(NULL, NULL, &opts);
    } else if (!strcmp(battery_name, "full")) {
        battery_full(NULL, NULL, &opts);
    } else if (strlen(battery_name) > 2 &&
        battery_name[0] == 'f' && battery_name[1] == '=') {
        battery_file(battery_name + 2, NULL, NULL, &opts);
    } else {
        fprintf(stderr, "Information about battery %s is absent\n",
            battery_name);
        return 1;
    }
    return 0;
}


static void apply_filter(GeneratorInfo **gi, GeneratorFilter filter, GeneratorInfo *filter_gen)
{
    switch (filter) {
    case FILTER_REVERSE_BITS:
        *filter_gen = define_reversed_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "All tests will be run with the reverse bits order\n");
        break;

    case FILTER_INTERLEAVED32:
        *filter_gen = define_interleaved_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "All tests will be run with the interleaved\n");
        break;

    case FILTER_HIGH32:
        *filter_gen = define_high32_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "All tests will be applied to the higher 32 bits only\n");
        break;

    case FILTER_LOW32:
        *filter_gen = define_low32_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "All tests will be applied to the lower 32 bits only\n");
        break;

    case FILTER_UINT31:
        *filter_gen = define_uint31_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "The lower bit will be added automatically\n");
        break;

    case FILTER_UINT63:
        *filter_gen = define_uint63_generator(*gi);
        *gi = filter_gen;
        fprintf(stderr, "The lower bit will be added automatically\n");
        break;

    case FILTER_NONE:
    case FILTER_UNKNOWN:
        break;
    }
}


static void print_ram_size(const CallerAPI *intf)
{
    RamInfo info;
    const long long mib_nbytes = 1ull << 20;
    const int ans = intf->get_ram_info(&info);
    const long long ramsize = info.phys_total_nbytes;
    if (ans == 0 || ramsize == RAM_SIZE_UNKNOWN) {
        intf->printf("Available/free RAM: unknown\n");
    } else if (ramsize > 64 * mib_nbytes) {
        intf->printf("Total physical RAM: %lld MiB\n", info.phys_total_nbytes / mib_nbytes);
        intf->printf("Free physical RAM:  %lld MiB\n", info.phys_avail_nbytes / mib_nbytes);
    } else {
        intf->printf("Total physical RAM: %lld KiB\n", info.phys_total_nbytes / 1024);
        intf->printf("Free  physical RAM: %lld KiB\n", info.phys_avail_nbytes / 1024);
    }
}


int main(int argc, char *argv[])
{
    char *battery_name, *generator_lib;
    if (argc < 2) {
        print_help();
        return 0;
    } else if (argc < 3) {
        if (!strcmp(argv[1], "--version")) {
            printf("SmokeRand %s\n", SMOKERAND_VERSION_FULL);
            return 0;
        } else {
            battery_name = "default";
            generator_lib = argv[1];
        }
    } else {
        // Check if battery name is omitted
        if (strlen(argv[2]) >= 2 && argv[2][0] == '-' && argv[2][0] == '-') {
            // Battery name is omitted: use `default`
            battery_name = "default";
            generator_lib = argv[1];
            argc++; // New offsets for `SmokeRandSettings_load`
            argv--;
        } else {
            // Battery name is given
            battery_name = argv[1];
            generator_lib = argv[2];
        }
    }

    SmokeRandSettings opts;
    if (SmokeRandSettings_load(&opts, argc, argv)) {
        return BATTERY_ERROR;
    }
    const int is_stdin32 = !strcmp(generator_lib, "stdin32");
    const int is_stdin64 = !strcmp(generator_lib, "stdin64");
    const int is_stdout =
        !strcmp(battery_name, "stdout") ||
        !strcmp(battery_name, "stdoutfl") ||
        !strcmp(battery_name, "stdoutflx") ;
    set_use_stderr_for_printf(is_stdout); // Messages mustn't be in PRNG output

    if (opts.nthreads > 1 && (is_stdin32 || is_stdin64)) {
        fprintf(stderr, "Multithreading is not supported for stdin32/stdin64\n");
        return BATTERY_ERROR;
    }

    if (!entfuncs_test()) {
        fprintf(stderr, "Seed generator self-test failed\n");
        return BATTERY_ERROR;
    }

    if (!strcmp(generator_lib, "list")) {
        return print_battery_info(battery_name);
    }

    if (is_stdin32 || is_stdin64) {
        CallerAPI intf = CallerAPI_init();
        GeneratorInfo stdin_gi;
        if (is_stdin32) {
            stdin_gi = StdinCollector_get_info(stdin_collector_32bit);
        } else {
            stdin_gi = StdinCollector_get_info(stdin_collector_64bit);
        }
        GeneratorInfo_print(&stdin_gi, is_stdout);
        BatteryExitCode ans = run_battery(battery_name, &stdin_gi, &intf, &opts);
        StdinCollector_print_report();
        CallerAPI_free();
        return ans;
    } else {
        GeneratorInfo filter_gen;
        CallerAPI intf = (opts.nthreads == 1) ? CallerAPI_init() : CallerAPI_init_mthr();
        GeneratorModule mod = GeneratorModule_load(generator_lib, &intf);
        if (!mod.valid) {
            CallerAPI_free();
            return BATTERY_ERROR;
        }
        GeneratorInfo *gi = &mod.gen;
        if (gi->nbits != 32 && opts.filter == FILTER_UINT31) {
            fprintf(stderr, "This filter is supported only for 32-bit generators\n");
            CallerAPI_free();
            return BATTERY_ERROR;
        }
        if (gi->nbits != 64 && opts.filter == FILTER_UINT63) {
            fprintf(stderr, "This filter is supported only for 64-bit generators\n");
            CallerAPI_free();
            return BATTERY_ERROR;
        }
        if (gi->nbits != 64 && (opts.filter == FILTER_INTERLEAVED32 ||
            opts.filter == FILTER_HIGH32 || opts.filter == FILTER_LOW32)) {
            fprintf(stderr, "This filter is supported only for 64-bit generators\n");
            CallerAPI_free();
            return BATTERY_ERROR;
        }
        apply_filter(&gi, opts.filter, &filter_gen);
        intf.printf("SmokeRand %s\n", SMOKERAND_VERSION_FULL);
        GeneratorInfo_print(gi, is_stdout);
        print_ram_size(&intf);
        BatteryExitCode ans = run_battery(battery_name, gi, &intf, &opts);
        GeneratorModule_unload(&mod);
        CallerAPI_free();
        return ans;
    }
}
